File: src/main/java/dev/emortal/nbstom/NBS.java
package dev.emortal.nbstom;

import net.kyori.adventure.audience.Audience;
import net.kyori.adventure.sound.Sound;
import net.minestom.server.entity.Player;
import net.minestom.server.timer.Scheduler;
import net.minestom.server.timer.Task;
import net.minestom.server.timer.TaskSchedule;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Supplier;

public class NBS {

    public static final Map<UUID, Task> playingTaskMap = new ConcurrentHashMap<>();

    /**
     * Plays this NBS song to an audience
     * Can be cancelled by using {@link #stop(UUID)} with the player's UUID. This stops automatically when the player leaves.
     *
     * @param player The player to play the song to
     */
    public static void play(NBSSong song, Player player) {
        play(song, player, player.scheduler(), player.getUuid());
    }

    /**
     * Plays this NBS song to an audience
     * Can be cancelled by using {@link #stop(UUID)} with the same stopId or by cancelling via the scheduler
     *
     * @param audience The audience to play the song to
     * @param scheduler The scheduler to tick the song on
     * @param stopId The id for use with {@link #stop(UUID)} later
     */
    public static void play(NBSSong song, Audience audience, Scheduler scheduler, UUID stopId) {
        playingTaskMap.put(stopId, scheduler.submitTask(new Supplier<>() {
            int tick = 0;

            @Override
            public TaskSchedule get() {
                if (tick > song.getLength() + 1) {
                    return TaskSchedule.stop();
                }

                List<Sound> sounds = song.getTicks().get(tick);
                if (sounds != null) {
                    for (Sound sound : sounds) {
                        audience.playSound(sound, Sound.Emitter.self());
                    }
                }

                tick++;

                return TaskSchedule.millis((long) (1000.0 / song.getTps()));
            }
        }));
    }

    public static void stop(Player player) {
        stop(player.getUuid());
    }
    public static void stop(UUID stopId) {
        Task task = playingTaskMap.get(stopId);
        if (task != null) {
            task.cancel();
        }
        playingTaskMap.remove(stopId);
    }

}

File: src/main/kotlin/dev/emortal/nbstom/NBS.kt
package dev.emortal.nbstom

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import net.kyori.adventure.sound.Sound
import net.minestom.server.entity.Player
import net.minestom.server.particle.Particle
import net.minestom.server.particle.ParticleCreator
import net.minestom.server.utils.PacketUtils
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Path
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ThreadLocalRandom
import kotlin.io.path.readBytes

/**
 * Class to read and play .nbs files
 */
class NBS(path: Path) {

    val version: Byte
    val instrumentCount: Byte
    val length: Int
    val layerCount: Int
    val songName: String
    val author: String
    val originalAuthor: String
    val description: String
    val tps: Double
    val autoSaving: Boolean
    val autoSavingDuration: Byte
    val timeSignature: Byte
    val minutesSpent: Int
    val leftClicks: Int
    val rightClicks: Int
    val noteBlocksAdded: Int
    val noteBlocksRemoved: Int
    val midiFileName: String
    val loop: Boolean
    val maxLoopCount: Byte
    val loopStartTick: Int

    val ticks: Array<NBSTick?>

    init {
        val bytes = path.readBytes()
        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)

        buffer.unsignedShort
        version = buffer.get()
        instrumentCount = buffer.get()
        length = buffer.unsignedShort
        layerCount = buffer.unsignedShort
        songName = buffer.getNBSString()
        author = buffer.getNBSString()
        originalAuthor = buffer.getNBSString()
        description = buffer.getNBSString()
        tps = buffer.unsignedShort / 100.0
        autoSaving = buffer.get().toInt() == 1
        autoSavingDuration = buffer.get()
        timeSignature = buffer.get()
        minutesSpent = buffer.int
        leftClicks = buffer.int
        rightClicks = buffer.int
        noteBlocksAdded = buffer.int
        noteBlocksRemoved = buffer.int
        midiFileName = buffer.getNBSString()
        loop = buffer.get().toInt() == 1
        maxLoopCount = buffer.get()
        loopStartTick = buffer.unsignedShort

        ticks = readNotes(buffer)
    }

    private fun readNotes(buffer: ByteBuffer): Array<NBSTick?> {
        var currentTick = -1
        val ticks = Array<NBSTick?>(length + 1) { null }

        while (true) {
            val tickJumps = buffer.unsignedShort
            if (tickJumps == 0) break
            currentTick += tickJumps

            val notes = readTickNoteLayers(buffer)
            ticks[currentTick] = NBSTick(currentTick, notes)
        }

        return ticks
    }

    private fun readTickNoteLayers(buffer: ByteBuffer): List<NBSNote> {
        val notes = mutableListOf<NBSNote>()
        var currentLayer = -1

        while (true) {
            val layerJumps = buffer.unsignedShort
            if (layerJumps == 0) break
            currentLayer += layerJumps

            notes.add(NBSNote.readNote(buffer))
        }

        return notes
    }


    companion object {
        private val playingTasks = ConcurrentHashMap<Player, MinestomRunnable>()

        private val scope = CoroutineScope(Dispatchers.IO)

        /**
         * Stops playing the song to a player
         * @param player The player to stop playing to
         */
        fun stopPlaying(player: Player) {
            playingTasks[player]?.cancel()
            playingTasks.remove(player)
        }

        /**
         * Play the song to a player
         * @param song The song to play
         * @param player The player to play the song to
         */
        fun play(song: NBS, player: Player) {
            val task = object : MinestomRunnable(repeat = Duration.ofMillis((1000.0 / song.tps).toLong()), iterations = song.length, coroutineScope = scope) {
                override suspend fun run() {
                    val nbstick = song.ticks[currentIteration.get() - 1]
                    nbstick?.notes?.forEach {
                        val sound = NBSNote.toSound(it)
                        player.playSound(sound, Sound.Emitter.self())
                    }
                }

            }

            playingTasks[player] = task
        }

        /**
         * Play the song to a player
         * @param song The song to play
         * @param player The player to play the song to
         */
        fun playWithParticles(song: NBS, player: Player, viewersToo: Boolean = true) {
            val task = object : MinestomRunnable(repeat = Duration.ofMillis((1000.0 / song.tps).toLong()), iterations = song.length, coroutineScope = scope) {
                val rand = ThreadLocalRandom.current()

                override suspend fun run() {
                    val nbstick = song.ticks[currentIteration.get() - 1]
                    nbstick?.notes?.forEach {
                        val sound = NBSNote.toSound(it)
                        player.playSound(sound, Sound.Emitter.self())

                        val useCount = it.key - 33
                        val particlePacket = ParticleCreator.createParticlePacket(Particle.NOTE, player.position.x + rand.nextDouble(-0.4, 0.4), player.position.y + 1.5 + rand.nextDouble(0.0, 0.2), player.position.z + rand.nextDouble(-0.4, 0.4), useCount / 24f, 0f, 0f, 1)

                        if (viewersToo) {
                            player.sendPacketToViewersAndSelf(particlePacket)
                        } else {
                            player.sendPacket(particlePacket)
                        }
                    }
                }

            }

            playingTasks[player] = task
        }
    }

}